src/ansys/simai/core/utils/auth.py

```diff
@@ -119,54 +148,59 @@ def from_request_device_auth(
                 },
             )
             if b"authorization_pending" not in validation.content:
-                return cls(**handle_response(validation), cache_file=cache_file)
+                return _AuthTokens(**handle_response(validation))
             # polling can't be faster or auth server returns HTTP 400 Slow down
             time.sleep(5)
 
-    def is_token_expired(self):
-        return datetime.utcnow() > (self.expiration - self._EXPIRATION_BUFFER)
-
-    def is_refresh_token_expired(self):
-        return datetime.utcnow() > (self.refresh_expiration - self._EXPIRATION_BUFFER)
-
-    def refresh(self, session: requests.Session, token_url: str) -> "_AuthTokens":
+    def _refresh_auth_token(self, refresh_token: str) -> Optional[_AuthTokens]:
         logger.debug("Refreshing authentication tokens.")
         request_params = {
             "client_id": "sdk",
             "grant_type": "refresh_token",
-            "refresh_token": self.refresh_token,
+            "refresh_token": refresh_token,
         }
         try:
             return _AuthTokens(
-                **handle_response(session.post(token_url, data=request_params)),
-                cache_file=self.cache_file,
+                **handle_response(self.session.post(self.token_url, data=request_params))
             )
-        except requests.exceptions.ConnectionError as e:
-            raise ConnectionError(e) from None
-
-
-def _get_cached_or_request_device_auth(
-    session: requests.Session,
-    device_auth_url: str,
-    token_url: str,
-    auth_cache_hash: Optional[str],
-):
-    auth = None
-    auth_cache_path = None
-    if auth_cache_hash:  # Try to fetch auth from cache
-        auth_cache_path = str(get_cache_dir() / f"tokens-{auth_cache_hash}.json")
-        auth = _AuthTokens.from_cache(auth_cache_path)
+        except (requests.exceptions.ConnectionError, ApiClientError) as e:
+            logger.error(f"Could not refresh authentication tokens: {e}")
+            return None
+
+    def _schedule_auth_refresh(self, refresh_expires_in: float):
+        """Schedule authentication refresh to avoids refresh token expiring if the client is idle for a long time."""
+        if refresh_expires_in <= self.REFRESH_BUFFER:
+            # if the refresh token expires this soon it means we are reaching max sso session lifespan
+            return
+        self.refresh_timer.cancel()
+        self.refresh_timer = threading.Timer(
+            refresh_expires_in - self.REFRESH_BUFFER, self.get_tokens
+        )
+        self.refresh_timer.daemon = True
+        self.refresh_timer.start()
```

Comment:
it looks like there might be a chance for a race condition here; 

if `refresh_expires_in - self.REFRESH_BUFFER` is small enough, let's say _0.1s_, and the execution of `get_tokens` lasts longer than _0.1s_, more than one instances of `get_tokens` would run in parallel. Since `get_tokens` reads/writes from/to file, this could create problems. WDYT?

Reply:
Yes I just reduced the odds but not eradicated the race condition to avoid locking a cross-process mutex for every request.
Fixed by 58dc4f0674066806c623a14dd47b1a5bc2ae3818

Reply:
seems reasonable 👌